python/cudaq/kernel/kernel_builder.py

```diff
@@ -231,6 +236,23 @@ def __init__(self, argTypeList):
 
             self.insertPoint = InsertionPoint.at_block_begin(e)
 
+    def __processArgType(self, ty):
+        """
+        Process input argument type. Specifically, try to infer the 
+        element type for a list, e.g. list[float]. 
+        """
+        if ty in [cudaq_runtime.qvector, cudaq_runtime.qubit]:
+            return ty, None
+        if typing.get_origin(ty) == list or isinstance(ty(), list):
+            if '[' in str(ty) and ']' in str(ty):
+                # Infer the slice type
+                result = re.search(r'ist\[(.*)\]', str(ty))
```

Comment:
I was initially concerned that this might not scale to a 2D list, but I think that wouldn't work anyway (for other reasons). From a language specification perspective, do we intend to allow it?

Reply:
Yea I played around with that a bit. I fell back to retaining the support we already had for the Python builder - we didn't allow `list[list[...]]` before so I think its fine to throw an error here for that. I think we can vet that decision in the language meetings and then adjust this in the future if necessary.